server/lib/scheduler/RootParentFinder.ts

```diff
@@ -0,0 +1,47 @@
+import * as models from "../models";
+
+class RootParentFinder {
+  private static RootParentCache = {};
+
+  public static clearRootParentCache() {
+    this.rootParentCache = {};
+  }
+
+  public static async getRootParentOfWorkspace(workspaceId, workspace) {
+    if (this.rootParentCache[workspaceId]) {
+      return this.rootParentCache[workspaceId];
+    }
+
+    if (!workspace) {
+      workspace = await models.Workspace.findById(workspaceId);
+    }
+
+    if (!workspace.parentId) {
+      this.rootParentCache[workspaceId] = workspace;
+      return workspace;
+    } else {
+      const rootParent = await this.getRootParentOfWorkspace(workspace.parentId);
+      this.rootParentCache[workspaceId] = rootParent;
+      return rootParent;
+    }
+  }
+
+  public static async getRootParentIdOfWorkspace(workspaceId) {
+    if (this.rootParentCache[workspaceId]) {
+      return this.rootParentCache[workspaceId].id;
+    }
+
+    const workspace = await models.Workspace.findById(workspaceId);
+
+    if (!workspace.parentId) {
+      this.rootParentCache[workspaceId] = workspace;
+      return workspace.id;
+    } else {
+      const rootParent = await this.getRootParentOfWorkspace(workspace.parentId);
+      this.rootParentCache[workspaceId] = rootParent;
+      return rootParent.id;
```

Comment:
This functionality seems very generic, like something I would expect in the workspace model file.

Reply:
As you might have guessed, this is here just because caching these "root parent" results is required for the scheduler to work. Without the caching here, performance is prohibitively bad.

Reply:
Not sure if it makes sense to put caching on the model.